### ISSUE TITLE
Review/feedback

### ISSUE THREAD

**Issue by poplexity** [cryptosuviio/tlos-recovery#1]
For the add action, we should check that we have not already called add with this account by checking the unstaking and recovering tables for it.

For unstake and recover, we should have an `uint8_t count` parameter, and then instead of doing one account per action call, we can do up to the `count` number of accounts per action call.  There will be a max per transaction based on the 30ms cap, but we can certainly do more than one per action.  I've used this pattern (https://github.com/Telos-Foundation/eosio.contracts/blob/75b0fa3b2706686def1b6daf6b2452a82a2ebfe2/telos.snaps/src/telos.snaps.cpp#L70) for deleting a lot of rows at a time where I don't know how many I can do at once when I write the contract.

Have you tested the transfer part?  I don't know if a privileged account can just transfer from another account like that... the transfer action does a `require_auth( from )` which I think will fail?  Otherwise what is `eosio.wrap` contract for if the `eosio` account can just impersonate?  I have not tested to confirm it won't work, but figured I would point this out.

**Comment by villesundell** [comment]
Thanks for taking a look :slightly_smiling_face: 

Separate checks are not needed for add(), because accounts are defined as primary keys, therefore cannot be added twice (and the check would be identical to the one already performed by emplace()).

Actually, I made the first implementation few hours before your message (9bb421447cb93c85491460f4b2989315c6484c44) but didn't push :slightly_smiling_face: I made a lot of improvements to the current master branch, though.

The transfer part works, because "privileged" flag gives the contract godlike superpowers, disabling auth checks for every action made by the contract.

(closing this, but feel free to open if you have more questions)

**Comment by villesundell** [comment]
With add(), there is one known corner case though: if account is added while staked, and then the account unstakes by themselves, and the contract operator (probably me) adds the account again, then unstake() would not proceed until the account is removed (since it's already in recover table, where it tries to re-add the account). Contract operator can freely remove accounts. Also the account can remove themselves.

However, this is not how the contract should be used, since it's vital that the accounts are added before contract is privileged, otherwise there is a huge security problem. So, the contract must be privileged only after adding is complete.
